### PR TITLE
Add primitives for controlling reduction

### PR DESCRIPTION
**Pull request by rlepigre**
TODO:
- [ ] `to_constr` may miss `unblock`/`block` because of the shortcuts for identity substitutions/liftings. It calls `Vars.subst_instance_constr` which doesn't have any code for finding our primitives.

Version 2 (new)
---------------------

This version relies on the following primitives, the main difference with the previous version being that:
- `run` (previously called `let_lazy`) now takes a `Blocked T` instead of a `T` as first argument,
- and `run` also does not have a dependent type anymore currently (its type uses a primitive type now, so there is a bit of code to write in primitive support to handle this).
```coq
Primitive Blocked := #blocked_type.

Primitive block : forall (T : Type), T -> Blocked T := #block.
Primitive unblock : forall (T : Type), Blocked T -> T := #unblock.

Primitive run : forall (T K : Type), Blocked T -> (T -> K) -> K := #run.

Arguments block {_} _.
Arguments unblock {_} _.
Arguments run {_ _} _ _.
```
What really changed since the previous version is the semantics of the primitives, which intuitively is now:
```
  e  --{full reduction with all flags}--> block v
 ------------------------------------------------- (f and e not reduced prior to running the primitive)
              run e f  --{whnf}->  f v

  e  --{full reduction with all flags}--> v        v not "block _"
 ------------------------------------------------------------------ (f and e not reduced prior to running the primitive)
                  run e f  --{whnf}->  run v f

  e  --{full reduction with all flags}--> block v
 ------------------------------------------------- (e not reduced prior to running the primitive)
            unblock e  --{whnf}->  v
            
  e  --{full reduction with all flags}--> v        v not "block _"
 ------------------------------------------------------------------ (e not reduced prior to running the primitive)
                unblock e  --{whnf}->  unblock v
            
  e  --{full reduction witout flags}--> v
 ----------------------------------------- (e not reduced prior to running the primitive)
       block e  --{whnf}->  block v
```
In the above, it is assumed that full reduction (whatever the flags) is implemented by first deducing to weak-head normal form, and then recursively fully evaluating under the head (and its arguments if any).

Basically, this allows for the following behaviour:
```
block (unblock (let n := 1 + 1 in block (n + n)))   --{whnf}--> block (2 + 2).
```
Here, even inside a `block`, code that is `unblock`-ed is fully evaluated until a `block` is reached. So, evaluating this expression proceeds by fully evaluating `unblock (let n := 1 + 1 in block (n + n))` with empty flags (only `block` and `unblock` being transparent). Since the head is `unblock`, its argument `let n := 1 + 1 in block (n + n)` is then fully evaluated with full flags, so this proceeds by extending the ambient substitution with a binding `(n := 1 + 1)`, additionally marked as coming from a "forced" context. Then, the term `block (n + n)` is reached, and its evaluation proceeds again by fully evaluating `n + n` with empty flags, which traverses the term and substitutes `n` with the fully evaluated form of `1 + 1` (since the value of `n` was marked as coming form a forced context in the substitution). So, now we have `unblock (block (2 + 2))`, and it is reduced to just `2 + 2` according the the reduction rule of `unblock`. Hence, we finally end up with `block (2 + 2)`, where the `block` is the outer one from the term we started with.

Version 1 (old)
-------------------

(The initial design was discussed with @ppedrot, @Janno and @gmalecha, and later also during the [Coq Call on 11/07/2023](https://github.com/coq/coq/wiki/Coq-Call-2023-07-11).)

**Edited with further documentation.**

This PR adds new primitives that are intended to be used to get finer-grained control on kernel reduction.
```coq
Primitive let_lazy : forall (T : Type) (K : T -> Type) (e : T), (forall v : T, K v) -> K e := #let_lazy.

Primitive Blocked := #blocked_type.
Primitive block : forall (T : Type), T -> Blocked T := #block.
Primitive unblock : forall (T : Type), Blocked T -> T := #unblock.

Arguments let_lazy {_ _} _ _.
Arguments block {_} _.
Arguments unblock {_} _.
```

The reduction rules for the primitives are the following:
```
  e  --{lazy}-->  v
 -------------------------
  let_lazy e f  --->  f v       (the arguments of [let_lazy] are not reduced before running the primitive)

 ----------------------------
  unblock (block e)  --->  e    (the argument of [unblock] is reduced before running the primitive)

 --------------
  block e -/->                  (the argument of [block] is not reduced, [block e] is stuck)
```

With these rules, we have the following properties:
- For all `T`, the canonical forms for type `Blocked T` are of the form `@block T e`, where `e` is any term.
- A term of the form `let_lazy f e` is definitionally equal to `f e`. The reason this is true is that `lazy` is compatible with the reduction engine used by weak-head normalization and conversion, and so `f v` remains convertible to `f e`.

**Note:** nothing here is specific to `lazy`, and it could be implemented for any reduction strategy, provided it is compatible with conversion / weak-head normalization. We went with `lazy` since it is already fully in the kernel.

**Note:** as noted by @SkySkimmer, we could implement `let_lazy` via a flag on the let-binding constructor in the AST. Although that option would definitely work, the change would be somewhat more invasive.

### CODE REVIEW

**Comment by ejgallego** [comment]
How does this compare to the existing `lock/unlock` mechanism?

**Comment by SkySkimmer** [comment]
Did you consider putting a flag on letin or lambda instead of using a constant? The difference would be for lambda `strict_fun x => body` (such that when applied the argument is reduced before substituting) the codomain type doesn't need to be repeated, and for letin the type doesn't need to be repeated and the bound term is used in the typing rule.

In general I imagine haskellers know more than us about what's useful to control call-by-need ish reductions even if they don't care about conversion, maybe we should get some contacts.

>However, the term let_lazy T K e f is reduced by first normalizing e using lazy to get some v

AFAICT the code is doing full lazy, not weak head. What drives this choice?

>The plan for (2) is to add primitives:

What do these do?

**Comment by Janno** [comment]
Another very important motivation for this change is the never-ending discussion around controlling reduction in reflective tactics (in general, not just in the case of telescopes). 

Consider the metaprogramming Rosetta Stone that was started during the CUDW this year. Even the most trivial reflective tactic for simplifying real arithmetic needs to maintain a whitelist of functions to reduce. The list isn't huge in that example (https://github.com/coq-community/metaprogramming-rosetta-stone/blob/main/real_simplifier/ltac2_reflexive/theories/RealSimpl.v#L471) but any change to the basic definitions (e.g. any kind of refactoring that introduces aliases) will cause the list to require changes.

Even if all definitions involved are considered stable and will never change I still think that the necessity of these whitelist is antithetical to modularity. Why should clients of my module need to know about a helper function that I used to define, say, exponentiation?

With the features in this MR (`let_lazy` and support for blocking reduction) this issue goes away entirely. To demonstrate, here's a sketch of how to adapt that Rosetta Stone example above to these new mechanisms:
1. The leaf nodes of the reified arithmetic expressions will be wrapped in applications of the `block`  primitive to safeguard uninterpreted/unknown terms from reduction.
2. The simplification function can safely be reduced using `lazy` (without a whitelist!) in the tactic and `let_lazy` in the proof term
3. The resulting simplified term is reflected back into `R` as before except for leaf nodes which get an extra `unblock` to allow these terms to unify with their original form (which doesn't have `block` and `unblock`).

As somebody who writes reflective tactics for a living (and, time permitting, for fun) I find this extremely compelling.

**Comment by herbelin** [comment]
So, the point is to statically attach a computational behavior to redexes of a term, here let-ins or some constants at least, vs choosing dynamically (using cbn, cbv, lazy and their customizations, in particular here using white lists) which computational behavior to apply to the let-ins or constants of this term?

**Comment by Janno** [comment]
> So, the point of (1) is to statically attach a computational behavior to redexes of a term, here let-ins or some constants at least, vs choosing dynamically (using cbn, cbv, lazy and their customizations, in particular here using white lists) which computational behavior to apply to the let-ins or constants of this term?

I am not sure I understand the question so I'll try to offer my perspective: The point of (1) is to provide a way to force a term to be fully reduced up to (2), i.e. up to blocked terms. AFAICT there is no comparable mechanism in Coq today. We can force terms that inhabit inductive types to weak head normal form by using `match`es and writing everything in CPS or we can use dynamic reduction tactics to fully reduce terms up to reduction flags but those dynamic reductions do not offer the possibility of blocking entire subterms from being reduced (or even inspected).

Additionally, dynamic reductions are (usually) recorded in the proof term using casts so already we are talking about possibly repeating a lot of (sub)terms in the proof term. Add to that the fact that the strategy itself (including potentially used whitelists) is not recorded in the term which means that the kernel might take a very different and possibly very inefficient route at `Qed` time.

**Comment by herbelin** [comment]
I can imagine the point of (2) (which would be a "native" form of lock/unlock??). What would be the rules btw? That `unblock (block t)` reduces into `t` and that `block t` is otherwise recognized by the different reduction strategies as freezing `t`?

Otherwise, is that correct that what the PR/MR currently does is to provide a new binary node `f t` (named `let_lazy`) such that each reduction mechanism (cbn, cbv, ...) can specify (with the `eval_lazy` field of a functor) a way to reduce `t` when using this strategy, with an intended meaning that the strategies that are clients of this special application node call `lazy` to reduce `t`?

**Comment by Janno** [comment]
> I can imagine the point of (2) (which would be a "native" form of lock/unlock??). What would be the rules btw? That unblock (block t) reduces into t and that block t is otherwise recognized by the different reduction strategies as freezing t?

That is correct.

> Otherwise, is that correct that what the PR/MR currently does is to provide a new binary node f t (named let_lazy) such that each reduction mechanism (cbn, cbv, ...) can specify (with the eval_lazy field of a functor) a way to reduce t when using this strategy, with an intended meaning that the strategies that are clients of this special application node call lazy to reduce t?

In the current implementation the ambient reduction is used for the argument of `let_lazy`. But this isn't so much a deliberate choice. Rather, I think it was the easiest to implement. Also, not all reduction strategies are currently supported. 

Generally, I think there is some uncertainty around the exact behavior of strategies other than `lazy` on `let_lazy`. In a model where `let_lazy` is only ever inserted into proof terms in a very fine-grained and manual fashion (i.e. not as part of some function) it doesn't really matter all that much since `let_lazy` will never appear in goals. If `let_lazy` is used in functions on which one needs to prove properties the picture changes. I think this will need to be pondered and discussed thoroughly. My current thought is that re-using the ambient reduction is the right call since some reduction strategies really ought not to reduce "too much" (whatever that means in the current proof) so calling `lazy`  might go counter to that expectation.

File: test.v
Comment (SkySkimmer) [inline test.v:1]: rename this file test-suite/success/force.v or some such thing

**Comment by SkySkimmer** [comment]
What is 
~~~
Eval lazy in ((unblock ((fun _ => block 42) tt))).
~~~
supposed to produce? I would think `42` but I get no reduction with your branch.

File: kernel/cClosure.ml [hunk 1]
```diff
@@ -1034,6 +1031,35 @@ module FNativeEntries =
       | FArray (_u,t,_ty) -> t
       | _ -> assert false
 
+    let get_blocked env () e =
```
Comment (SkySkimmer) [inline kernel/cClosure.ml:1034]: This implementation looks wrong to me. I think `unblock` should specify the second argument as needing to be whnf, then you can just look for FApp of `block`.

Comment (rlepigre) [reply kernel/cClosure.ml]: I pushed a fix for this. I originally tried avoiding changing the code too much, but it is a bit hard to avoid adding a new `fterm` constructor for primitives that are values, like `block e`. What I have now works fine in the kernel (see 22936de55a6c8f8f753fc3ecefcb232868d0431d), but cannot be easily replicated in reduction machines based on `Constr.t`.

Morally, we should add a `Blocked of 'types * 'constr` to the `Constr.kind_of_term` type, but that sounds annoying. A better option would be to instead have a new type in `kernel/cPrimitives.ml` for "primitive type constructors", which would also encompass primitive integers, floats and arrays. So, something like:
```ocaml
type ('constr, 'types, 'sort, 'univs) prim_const =
  | Int of Uint63.t
  | Float of Float64.t
  | Array of 'univs * 'constr array * 'constr * 'types
  | Block of 'types * 'constr
```
And then, we would modify the type `Constr.kind_of_term` in the following way:
```diff
-   | Int       of Uint63.t
-   | Float     of Float64.t
-   | Array     of 'univs * 'constr array * 'constr * 'types
+   | PConst    of ('constr, 'types, 'sort, 'univs) prim_const
```

In general, the code around primitives is quite fragile: on master, primitives without any argument of primitive types lead to an assert failure. I fixed a few things, in the MR, but it is unlikely that this is good enough, especially not in the long run.

Basically, I'm not quite sure how to proceed from here, so I'd appreciate some guidance @SkySkimmer @ppedrot.

Comment (rlepigre) [reply kernel/cClosure.ml]: Also, the idea with this new type I'm proposing, my idea is that primitives should be able to manipulate elements of this type parametrically over the type parameters, and that would all happen in the code for running primitives in `kernel/primred.ml`.

**Comment by SkySkimmer** [comment]
I'll probably wait until after holidays (a couple weeks) to investigate further

**Comment by JasonGross** [comment]
Some suggestions:
1. Why `let_lazy` instead of `red_full` as in

```coq
Primitive red_full : forall (T : Type), T -> T := #eager.
Arguments red_full {_} _.
```

```
  e  --{redexpr}-->*  v
 -------------------------
  red_full e  ->  v
```
That is, `red_full e` reduces in a single step to the normal form of `e` (including under binders, with fixpoint refolding if in simpl, etc) 

2. All constants should be universe polymorphic if they are not already
3. `unblock` should reduce `block e args...`, that is, `unblock ((block f) x)` should not be stuck but should reduce (maybe the code already does this, but the description should make it clear)
4. What should happen with `unblock 5`?  Presumably when there is no `block` and the term is normal, it should reduce?  Or is this deliberately stuck?  But if it's stuck, then `unblock (unblock 5)` should reduce, I think?
5. What should happen with `unblock (fun _ => block 5) tt`?  Perhaps the rule should be that if `unblock f` is stuck, then `unblock f x` reduces to `unblock (f x)`?
6. Is this at all related to / compatible with my desire to have a variant of zeta reduction that steps `f (let x := v in b)` to `let x := v in f b`?

**Comment by Janno** [comment]
We are still changing the design and haven't had time to write documentation for the latest proposal that we plan to implement soon. Nonetheless, I think I can answer all questions regardless of the exact design.

> Why `let_lazy` instead of `red_full` as in [..]

The name `let_lazy` is very, very confusing and we will definitely change it. The reason it is called `lazy` is that it's using strongest reduction available in the kernel: `lazy`. It is in fact eager.

> That is, red_full e reduces in a single step to the normal form of e (including under binders, with fixpoint refolding if in simpl, etc)

The bare minimum implementation of the feature is to support reduction only via `lazy` and at `Qed` time. Our current thinking is that we will need a bit of a refactoring to support other reduction strategies. And it is not entirely clear yet how `let_lazy` (or whatever it will be called) will behave in other reduction strategies. My guess is that it makes the most sense to simply use whatever reduction strategy was called. That way you would get refolding with `simpl` and all the other oddities of the different reduction strategies.

> All constants should be universe polymorphic if they are not already

Agreed. I hope the primitive mechanism allows for that. From what I remember it should be fine.

> unblock should reduce block e args..., that is, unblock ((block f) x) should not be stuck but should reduce (maybe the code already does this, but the description should make it clear)

I don't think any of our designs allow for that reduction step. In fact, for our own use case it is crucial that beta reduction is stuck in this case.

> What should happen with unblock 5? Presumably when there is no block and the term is normal, it should reduce? Or is this deliberately stuck? But if it's stuck, then unblock (unblock 5) should reduce, I think?

> What should happen with unblock (fun _ => block 5) tt? Perhaps the rule should be that if unblock f is stuck, then unblock f x reduces to unblock (f x)?

The types wouldn't allow for that. `unblock` has type `Blocked X -> X`.

> Is this at all related to / compatible with my desire to have a variant of zeta reduction that steps f (let x := v in b) to let x := v in f b?

I don't know that any reduction strategy currently implemented allows for that. Our intent is not to extend the set of available reductions but instead to make at least one eager reduction available at Qed time in a way that it can be triggered explicitly.

**Comment by SkySkimmer** [comment]
>if you only care about let_lazy it can be implemented in terms of red_full:

That's wrong, it will also reduce the type of `e` (which may be slow even when `e` is fast)

**Comment by JasonGross** [comment]
> > if you only care about let_lazy it can be implemented in terms of red_full:
> 
> That's wrong, it will also reduce the type of `e` (which may be slow even when `e` is fast)

Ah, I see.

**Comment by SkySkimmer** [comment]
OTOH red_full could be combined with a let_strict which only head reduces.

**Comment by rlepigre** [comment]
I just pushed a rebase/cleanup, nothing new besides `let_lazy` being renamed into `run`.

**Comment by JasonGross** [comment]
The reduction rules suggest that `fun x => unblock (expensive identity function applied to x)` is fully normalized, because the argument to `unblock` reduces to a binder which does not reduce to a term headed by `block`.  (And similarly for an axiom.) Is this true?  This seems like it might be expensive to deal with, if reduction will keep backtracking over attempts to reduce the argument?

**Comment by rlepigre** [comment]
@JasonGross this is indeed, true, but some rules are missing in the description. In particular, any progress made will be kept. So, the example you give will just be reduced to `fun x => unblock x` if you hit it with `Eval lazy` for example.

**Comment by Janno** [comment]
We are making good progress. The PR already contains examples that show much it improves the ergonomics of proofs by reflection (where it can completely remove the need for reduction whitelists). But so far we hadn't managed to show actual performance improvements. In fact, the performance with forcing was worse than that without. With help from @ppedrot we have now made progress on that aspect of the work. 

We still have not managed to substantially speed up our original telescope example but that's only due to a terrible hack that forces way too many `to_constr` computations. Nonetheless, we can still show more general performance improvements that can be had with forcing.

The example below shows how the forcing primitives can be used to avoid the loss of sharing that happens between a function type's argument type and the function type's body. I am not aware of any way to circumvent this loss of sharing on master.

```coq
Fixpoint prods (n : nat) (t b : Type) : Type := (* [n] many nested [forall _:T] ending in [b] *)
  match n with | O => b | S n => forall _ : t, prods n t b end.
Fixpoint slow (n:nat) : nat :=    (* [slow n] is [0] but takes a while to get there *)
  match n with | O => O | S n => slow n + slow n end.
Ltac tac n m F x :=               (* build and apply [F m n x .. {m times total} .. x] *)
  let rec go m t :=
    lazymatch m with | O => exact t | S ?m => go m uconstr:(t x) end
  in
  go m (F m n).

Module Base.
  Axiom F : forall m n, let k := slow n in prods m (k = 0) True.
  Goal True. tac 19  1 F constr:(eq_refl O). Time Qed. (*    X *)
  Goal True. tac 19  2 F constr:(eq_refl O). Time Qed. (*  2*X *)
  Goal True. tac 19  3 F constr:(eq_refl O). Time Qed. (*  3*X *)
  Goal True. tac 19 10 F constr:(eq_refl O). Time Qed. (* 10*X *)
End Base.

Module Forced.
  Require Import Coq.Force.Force.
  Axiom F : forall m n, run (let k := slow n in block k) (fun k => prods m (k = 0) True).
  Goal True. tac 19  1 F constr:(eq_refl O). Time Qed. (* X *)
  Goal True. tac 19  2 F constr:(eq_refl O). Time Qed. (* X *)
  Goal True. tac 19  3 F constr:(eq_refl O). Time Qed. (* X *)
  Goal True. tac 19 10 F constr:(eq_refl O). Time Qed. (* X *)
End Forced.
```
The code constructs a function type with many arguments that all re-use the same expensive computation that is let bound at the start of the type. Checking an application of an inhabitant of that type to arguments which satisfy the reduced type trivially (but not the unreduced one) demonstrates how the computation is not shared by default. However, with the forcing infrastructure in this PR we can restore sharing and thus avoid repeated work at every individual argument.

(The remaining work per argument, i.e. checking `eq_refl O` against `O = O`, is dwarfed by the time it takes to reduce `slow n` just once. This means the the last four `Time Qed` only take the same time in the sense that all measurements are dominated by the reduction of `slow n`.)

**Comment by Janno** [comment]
Can we run the benchmarks to see how bad the changes impact developments that don't use the new primitives? We had to disable at least one shortcut (the `FConstruct` case in `knr` now always needs to traverse the stack to figure out if there is an `ZCaseT` even when the match flag is locally disabled) and there is some new indirection through `knht_app` that could affect performance as well.

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
Restarted the bench as inria gitlab had some issues

**Comment by SkySkimmer** [comment]
@coqbot bench (let's see how much the fconstruct thing changes)

**Comment by Janno** [comment]
Thanks for starting a new benchmark! It's not quite done but it looks like the result is entirely unchanged. I wonder if I misread the `perf stat -x` output I used to count instructions? Even the timings on fingroup seemed to show an improvement although I guess that could have been due to noise on my machine.

I suppose I'll have to go to one of the files that show a big regression and perf individual statements.

**Comment by Janno** [comment]
I am focusing on https://coq.gitlabpages.inria.fr/-/coq/-/jobs/3456109/artifacts/_bench/html/coq-engine-bench-lite/coq/PerformanceDemos/one_step_reduction.v.html#L214 because it is very easy to compile and consistently slower with the code in this PR.

I locally see a very repeatable +3s difference (21s vs 24s while `perf`ing with 99 samples per second). I don't have great tools to perform differential performance analysis but I have uploaded individual profiles (the output of `perf script`) here:

1. (master, 967e2222e79def2c15c6652de2e9f4c015a6dc89) http://janno-kaiser.de/bench-master.script
2. (this PR, br/letlazy, b07942c5b23c424d0d8dbac958fdc377eeaac962)  http://janno-kaiser.de/bench-b079.script

Here's a firefox profiler view of both profiles:
1. (master) https://profiler.firefox.com/public/8p9amjbfyj2m2z4jgd4f7wzb0vwrde0y77cnq0g/flame-graph/?globalTrackOrder=0&thread=0&timelineType=category&v=10
2. (PR) https://profiler.firefox.com/public/yrtpwwvjfwmzdgcmb4g6ks2t8j98ram8kwgdq2r/flame-graph/?globalTrackOrder=0&thread=0&timelineType=category&v=10

I  just re-discovered www.speedscope.app which makes comparing things on a function-by-function basis a little easier (in separate tabs, though, since it still lacks native support for comparing profiles). This is what I am currently using to perform the comparison. Unfortunately it doesn't have a nice sharing function like profiler.firefox.com does.

I always run `Optimize Heap` before running the actual statements and managed to get quite similar profiles without too much of the usual GC noise. The main differences I see are the following:
1. `knr` +1.5s, 
   * 0.4s are spent in `knr` itself  - not sure why
   * 0.5s are spent in `strip_rec` which, for some reason, appears without its caller `strip_update_shift_app` in the master profile - also not sure why; there should not be more or different items on the stack
   * Not sure about the remaining 0.6s. They must be spread out over all the remaining functions called by `knr`.
2. `knht` -0.9
   * A negative delta is expected here since some of the work (the `App` case) is now done in `knht_app`
3. `knht_app` (new code) +2.4s
   * Because the new primitives behave differently when partially applied, we need to find applications of them before arguments are moved to the stack.
   * `knht_app` runs  `Table.lookup_primitive` which is `Table.lookup` restricted to `CPrimitive`s
   * This has the highest potential to repeat work since it performs some work for non-`CPrimitive`s but exits early and does not cache the result. The reason for exiting early in these cases is to avoid code duplication with `knr`'s `FFLex` case.
   * `knr` will repeat the lookup for non-`CPrimitive`s but I am unable to find that in the profile.
   * Since +2.4 is not the inverse of -0.9 (the `knht` speedup) I suspect this shows the true overhead of `knht_app` to be roughly +1.5s

In total we have 1.5 + 2.4 - 0.9 = 3s which is the total overhead. The differences in these three functions thus seem to capture the entire difference in that particular line.

I am not quite sure what to do about this. If the overhead truly is due to the repeated lookup work between `knht_app` and `knr` then maybe `knht_app` could go all the way and basically inline `knr`'s `FFLex` case? We'd need to run `knht_app` also for the `Constant` case to capture unapplied terms. I don't like that change very much but I also don't know what else we could do to win back performance.

**Comment by Janno** [comment]
Removing the call to `knht_app` and using the old code for `knht`'s `App` case saves 1.5s which seems to validate the numbers in the previous comment. 

Unfortunately `knr` is still exactly 1.5s slower and I am not sure how that could happen.

**Comment by SkySkimmer** [comment]
>0.5s are spent in strip_rec which, for some reason, appears without its caller strip_update_shift_app in the master profile - also not sure why; there should not be more or different items on the stack

Sounds like tail call optimization.
BTW how do you know it's the strip_update_shift_app strip_rec and not eg the compact_stack strip_rec?

**Comment by Janno** [comment]
> Sounds like tail call optimization.

Interesting. That made me check the diff again and we have indeed changed `strip_update_shift_app`. The underlying function `strip_update_shift_app_red`, previously discarded the updated `head` argument. Now it returns the updated term and `strip_update_shift_app` simply discards it. And of course that means it's no simply a tail call.. there's a `snd` after the call.

I'll restore the original function for all the cases that don't need the updated head term. Let's see if that helps with `knr` timings.

**Comment by Janno** [comment]
Restoring the original `strip_update_shift_app_red` for all the previously existing callers restores the tail call in `strip_update_shift_app` and removes the associated +0.5s overhead in `knr`. This leaves +1s in `knr`, of which there are still +0.4s is in `knr` itself and +0.6 spread across various functions inside. I can't really dig into it much more because at that level GC differences make the timing differences too noisy.

**Comment by Janno** [comment]
> What if you extend the source code examples from 15 to 20?

Thanks! This comment made me investigate what is actually happening in the example and it turns out that only those calls with a `true` argument actually do typechecking (because of `abstract`) and thus kernel reduction. I am now using 
`  idtac 5; test_slow_zeta1 true 5; test_fast_zeta1 true 5; test_slow_zeta2 true 5; test_fast_zeta2 true 5` and that gives me a test case which runs for about 50s in total, most of which is in the kernel.

**Comment by Janno** [comment]
It seems in this case more samples does not equal more insight. Even with all the modifications I mentioned earlier (bypassing `knht_app` in favor of the old behavior of `knht` on `App` nodes, restoring `strip_update_shift_app` to its old behavior) I see a difference of roughly 3s between master and my code (46s -> 49s). The difference is made up of:

1. `knr` +1.9 with +0.4 of that in `knr` itself and the rest apparently spread very evenly in all its children
2. `knht` +1.4, all of that in `knht` itself.

The +1.5s in `knr`'s children could be from garbage collection. `fconstr` is one field bigger in this PR.

Regarding the self time in `knr` and `knht`: I wonder if we are seeing the result of assembly changes because of the additional field in `fconstr`, the additional `~mode` argument in many functions, and the additional `tab` argument in some function. 

In any case, I do not know how to continue. Once @rlepigre returns we'll probably have more ideas on how to investigate and improve this.

**Comment by SkySkimmer** [comment]
>The +1.5s in knr's children could be from garbage collection. fconstr is one field bigger in this PR.

You could try bit twiddling to fix conv_or_red and mode in 1 int? We did that for some time for sprop marks (f3c1fe8bc7979cd7036fcd2a1b2fe8f0a1acb058) until we changed how the machine deals with irrelevance.

**Comment by rlepigre** [comment]
> >The +1.5s in knr's children could be from garbage collection. fconstr is one field bigger in this PR.
> 
> You could try bit twiddling to fix conv_or_red and mode in 1 int? We did that for some time for sprop marks (f3c1fe8bc7979cd7036fcd2a1b2fe8f0a1acb058) until we changed how the machine deals with irrelevance.

Yeah, that was my plan to eventually do that once things worked. Not sure that will buy us enough though.

**Comment by Janno** [comment]
I pushed the `strip_update_shift_app` change and a very naive version of packing `mode` and `mark` into a single integer. I haven't done any bit fiddling in OCaml before so my attempt might be suboptimal. It seems to help with timings locally but I've said that before and it didn't end up helping in the benchmark.

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by Janno** [comment]
`coq-engine-bench-lite` is a tiny bit worse than before. Next time I'll let the pros handle the bits...

**Comment by Janno** [comment]
@coqbot bench

**Comment by Janno** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
Something went wrong with the bench
It seems the commit for https://gitlab.inria.fr/coq/coq/-/pipelines/867401 isn't an auto merge commit from the bot 
Let's retry

@coqbot run ci

**Comment by SkySkimmer** [comment]
@coqbot run light ci

**Comment by SkySkimmer** [comment]
Seems like it's not just a problem here actually
https://github.com/coq/bot/issues/294

**Comment by Janno** [comment]
There's also an actual test failure in our tests so we should wait with the bench until we fix that.

**Comment by Janno** [comment]
@coqbot bench

**Comment by Janno** [comment]
@coqbot bench

**Comment by Janno** [comment]
@coqbot bench

**Comment by Janno** [comment]
@coqbot bench

**Comment by Janno** [comment]
@coqbot bench

**Comment by JasonGross** [comment]
Is there any hope (or theory) of extracting these to OCaml in a way that gives the extracted code correct reduction behavior?

(I'm trying to figure out how, given `(fun x => <some thunk>) : A -> B * C`, to get out `A -> B` and `A -> C` without having to run the code in the thunk twice.  (Maybe there's some way with ref cells or delimited non-linear continuations?))

**Comment by rlepigre** [comment]
> Is there any hope (or theory) of extracting these to OCaml in a way that gives the extracted code correct reduction behavior?
> 
> (I'm trying to figure out how, given `(fun x => <some thunk>) : A -> B * C`, to get out `A -> B` and `A -> C` without having to run the code in the thunk twice. (Maybe there's some way with ref cells or delimited non-linear continuations?))

I haven't given that much thought, but I find it unlikely that something like that would be possible. You can't really do any evaluation under closures in OCaml as far as I know, at least not dynamically.
